python/dgl/node_flow.py

```diff
@@ -1,54 +1,77 @@
 """Class for NodeFlow data structure."""
 from __future__ import absolute_import
 
+import ctypes
+
+from ._ffi.function import _init_api
 from .base import ALL, is_all, DGLError
 from . import backend as F
 from .frame import Frame, FrameRef
 from .graph import DGLBaseGraph
+from .graph_index import GraphIndex, transform_ids
 from .runtime import ir, scheduler, Runtime
 from . import utils
 from .view import LayerView, BlockView
 
-def _copy_to_like(arr1, arr2):
-    return F.copy_to(arr1, F.context(arr2))
+__all__ = ['NodeFlow']
 
-def _get_frame(frame, names, ids):
-    col_dict = {name: frame[name][_copy_to_like(ids, frame[name])] for name in names}
-    if len(col_dict) == 0:
-        return FrameRef(Frame(num_rows=len(ids)))
-    else:
-        return FrameRef(Frame(col_dict))
+NodeFlowHandle = ctypes.c_void_p
 
+class NodeFlowIndex(GraphIndex):
+    """NodeFlow index object.
 
-def _update_frame(frame, names, ids, new_frame):
-    col_dict = {name: new_frame[name] for name in names}
-    if len(col_dict) > 0:
-        frame.update_rows(ids, FrameRef(Frame(col_dict)), inplace=True)
+    Parameters
+    ----------
+    handle : NodeFlowHandle
+        Handler
+    """
+    def __init__(self, handle):
+        super(NodeFlowIndex, self).__init__(
+            handle, multigraph=False, readonly=True)
 
+    def __del__(self):
+        """Free this nodeflow index object."""
+        _CAPI_NodeFlowFree(self._handle)
```

Comment:
I see. Then I will put node_mapping, edge_mapping members back to NodeFlowIndex so they could be correctly pickled.

Reply:
Add tests for pickling. Ends up completely delete the NodeFlowIndex class (it is actually not used at all). Since all the members of `NodeFlow` is picklable, the `NodeFlow` object is also picklable and no set/get state functions need to be implemented.